### PR TITLE
Implements proper message routing logic

### PR DESCRIPTION
**Pull request by adamnfish**
The original logic was very simplified, and now we're actually integrating a service we see that it was well short of what is required. This PR introduces "proper" contact resolution, that behaves the way people expect.

This works using a hierarchy of targets. App matches are more important than Stack matches, which are more important that AwsAccount matches. Stage must always match and is assumed to be PROD if it is omitted.

If there's an exact match between targets and a mapping then life is easy, but that won't generally be the case. We want to be able to send a message to `App("myapp")`. We'd also like to be able to send a message from an application that derives all the targets from its tags (e.g. `App("myapp")`, `Stack("mystack")`, `AwsAccount("account-id")`, `Stage("STAGE")`). In both case it should route messages to the "correct place". The correct place would be an exact (/close) match if it exists in the configuration, but if it hasn't been fully specified we'd still expect it to be able to send a message to the best valid match (App if that exists, falling back to Stack and then AwsAccount).

It is pretty complicated, but the logic ends up being reasonable and there are **a lot** of tests to capture all the edge cases. It also has a few "integration" tests that simulate more realistic invocations against a more realistic mapping configuration.

Sorry this is so complex, and such a large PR. Maybe start with the integration test section of ContactsTest and then look at the implementation.

This logic basically *is* Anghammarad. Sending messages to a webhook is easy, the value of a department-wide service is its ability to send messages to the correct people.

### CODE REVIEW

File: anghammarad/src/main/scala/com/gu/anghammarad/Contacts.scala [hunk 1]
```diff
@@ -2,34 +2,114 @@ package com.gu.anghammarad
...
+  /**
+    * Searches among mappings that have more detail than requested.
+    *
+    * e.g.
+    * Ask for:
+    *   App("app")
+    * Mapping has:
+    *   Stack("stack"), App("app")
+    *
+    * This is tricky because we wouldn't to match on the following
```
Comment (katebee) [inline anghammarad/src/main/scala/com/gu/anghammarad/Contacts.scala]: > This is tricky because we wouldn't to match on the following

Missing a word? 😄 

> This is tricky because we wouldn't **want** to match on the following

Comment (adamnfish) [reply anghammarad/src/main/scala/com/gu/anghammarad/Contacts.scala]: Great spot, thank you for reading it so carefully :-)
